src/main/java/com/minecolonies/coremod/client/gui/WindowResearchTree.java

```diff
@@ -74,57 +99,94 @@ public WindowResearchTree(final String branch, final BuildingUniversity.View bui
 
         final ZoomDragView view = findPaneOfTypeByID(DRAG_VIEW_ID, ZoomDragView.class);
 
-        drawTree(0, 0, view, researchList, building.getColony().getResearchManager().getResearchTree(), true, false, 0);
-
-        for (int i = 1; i <= MAX_DEPTH; i++)
-        {
-            final Label timeLabel = new Label();
-            timeLabel.setLabelText(Math.pow(2, i - 1) + "h");
-            timeLabel.setPosition((i - 1) * (GRADIENT_WIDTH + X_SPACING) + GRADIENT_WIDTH / 2, TIMELABEL_Y_POSITION);
-            timeLabel.setColor(Color.rgbaToInt(218, 202, 171, 255));
-            view.addChild(timeLabel);
-        }
+        drawTreeBackground(view);
+        drawTree(0, 0, view, researchList, false);
     }
 
-    /**
-     *
-     * @param researchRequirement requirement to start research
-     * @return whether the requirement has been satisfied for research
-     */
-    private boolean isResearchFulfilled(final BuildingResearchRequirement researchRequirement)
+    @Override
+    public void onButtonClicked(@NotNull final Button button)
     {
-        List<IBuildingView> buildings = building.getColony().getBuildings();
-        int levels = 0;
-
-        if (researchRequirement == null)
+        super.onButtonClicked(button);
+        if (button.getParent().getChildren().contains(undoButton))
```

Comment:
not sure about this button handling, shouldn't be in "if undo pressed" branch?
and also the for cycle below in this if branch?

Reply:
These are a hacky way to handle removing the "undo" button, icon, and tooltips, and then replace the original contents of the tooltips they were taking the position of (bc multiple overlapping items with hovertext will render all of the hovertext in a mouse position), if any button other than the confirm-undo is pressed.  Because the positions are so malleable, it's hard to redraw individual Research Items without redrawing the entire branch, but I'm certainly open to different approaches.

It technically doesn't need to be done for logic paths where the research tree gets closed, but the performance cost of doing them is pretty low, and eventually we may want the ability to handle begin-research or cancel-research events without having to close the tree.

Reply:
> bc multiple overlapping items with hovertext will render all of the hovertext in a mouse position

how about making an issue and let someone actually fix it? :D

_btw towards "hacky ways", sure use them, but make sure they work under every condition and comment them_

Reply:
Fair.  I could see cases where it'd be useful and intentional -- the ItemIcons render their tooltips over the research tooltips, for example, which I kinda like -- but should have brought it up in Structurize.

Definitely will implement better comments for the hacky bits here and elsewhere, rather than just the javadocs.